### ISSUE TITLE
How to print web page using vivliostyle.js

### ISSUE THREAD

**Issue by steinerkelvin** [vivliostyle/vivliostyle.js#272]
I haven't found any docs so I would like to know how to include vivliostyle.js in a web page with some Paged Media CSS and print it.

**Comment by frivoal** [comment]
If you just want to print, the easiest is not to include vivliostyle.js in the webpage, and instead use [Vivliostyle Formatter](http://vivliostyle.com/en/products/#secProducts-viv-formatter) from the command line like this:

```
vivliostyle-formatter http://www.example.com/page/you/want/to/print/ -o filename.pdf
```

Another alternative that doesn't require installing vivliostyle.js in your website is to use the [chrome extension](https://chrome.google.com/webstore/detail/vivliostyle/ffeiildjegeigkbobbakjjmfeacadbne). Install it, load the page you want, click the vivliostyle icon, tick the paginate checkbox, click apply, then use chrome's normal print dialog.

If you install vivliostyle.js into your website, you will also be able to print from your browser's print dialog, but in addition, you will offer the site's visitors a paginated view directly in the browser. Here are a few samples of what it looks like: http://vivliostyle.com/en/samples/ (careful, some are heavy documents).

If that's what you want to do, instead of fetching vivliostyle.js from github, I suggest you use the packaged released version from here: http://vivliostyle.com/en/products/#secProducts-viv-js. After you download and unzip the file, there's documentation on how to install it inside it.

**Comment by steinerkelvin** [comment]
Hello, I've tried `vivliostyle-formatter` and the Chrome extension and they are very interesting. I'm pretty impressed with vivliostyle :D

I was also able to make the package to work on my server, and that is probably sufficient to my purpose, but I would like to know if I can run vivliostyle on some dynamically generated content (in the browser), weither by passing a string to the library or injecting vivliostyle in an iframe, like what the Chrome extension does.

I would also like to know if there are plans to implement CSS named pages.

Thank you very much  :)

**Comment by frivoal** [comment]
> I'm pretty impressed with vivliostyle :D

Happy to hear that :)

> I would like to know if I can run vivliostyle on some dynamically generated content (in the browser), weither by passing a string to the library or injecting vivliostyle in an iframe, like what the Chrome extension does.

Vivliostyle Viewer (i.e. the bundle that contains both [Vivliostyle.js](https://github.com/vivliostyle/vivliostyle.js) and [vivliostyle-ui](https://github.com/vivliostyle/vivliostyle.js) does not currently support that, although we do plan to add this in a future version.

The engine of Vivliostyle.js can do this, as you can see from the chrome extension, but then you need to write your own alternative to [vivliostyle-ui](https://github.com/vivliostyle/vivliostyle.js) to invoke it.

> I would also like to know if there are plans to implement CSS named pages.

We have not announced a date, but yes, we intend to support this. There are some complications because we also support page templates as defined in [Adaptive Layout](http://www.idpf.org/epub/pgt/), and it is not clear from any specification how the two are supposed to interact. But we'll get there. In the meanwhile, you may want to have a look at [Adaptive Layout](http://www.idpf.org/epub/pgt/). The syntax isn't always intuitive, but it is very powerful, and in many case, you can use it instead of named pages.

If you want to give me some details about what you're trying to do with named pages, I may be able to show you how to do the same using [Adaptive Layout](http://www.idpf.org/epub/pgt/).

**Comment by steinerkelvin** [comment]
I've read some docs, and the syntax of Adaptative Layout seems a bit complicated indeed. But I think I'll able to get what I need.
By now I am trying to repeat a header over all pages excluding the first one (like [these docs](http://sorotokin.com/adaptive-layout/tutorials/02-running-header/style.html)) and setting different margins for alternate pages (like what we can do using CSS Page Media [pseudo-classes  `:left` and `:right`](https://www.w3.org/TR/css3-page/#spread-pseudos).

Now I found a problem which I don't know if is a bug or it is me doing something wrong.
Basically, if I set a partition with margins the content overflow the borders, as we can see in this sample file:
https://gist.github.com/kelvinss/b40a8c16fbd5191557b69f225c5f2d04#file-screenshot-png
However, the page renders ok if I set any value to `column-width`, even if the value is greater than the page width.

**Comment by frivoal** [comment]
You should be able to fix your overflow bug by adding

```
right: 0;
left: 0;
```

after `margin: 2cm;` to make sure that the partition is forced to stay within the page bounds.

For repeat headers, have a look at `-epubx-flow-options: static` if you just want the same header to be repeated everywhere, or `-epubx-flow-options: static last` if you want it to change based on the content (for example, use the title of the current chapter).

For first, left and right, pages, you can use  `-epubx-enabled: -epubx-expr( page-number == 1 )`,  `-epubx-enabled: -epubx-expr( page-number % 2 == 1 )` or `-epubx-enabled: -epubx-expr( page-number % 2 == 0 )`
